### PR TITLE
Incremental Frequency and Resolution

### PR DESCRIPTION
**Pull request by Robertboy18**
I'm excited to share with you the final update for the Incremental version of Neural Operators. This pull request includes the latest changes and improvements to the Incremental module, bringing it to its completed state.

Here's a brief overview of the updates included in this pull request:

Algorithm 1: Incremental
Added a new method called loss_gap(loss) to handle the incremental mode increase based on the loss gap. The method tracks the loss values and increases the incremental modes if the difference between consecutive losses falls below a predefined threshold.

Algorithm 2: Gradient-based Explained Ratio
Implemented the grad_explained() method to handle the incremental mode increase based on the explained variance of gradients. The method accumulates gradients and checks the explained variance of each mode. The incremental mode is increased if the explained variance is below a specified threshold.

Algorithm 3: Regularize Input Resolution
Added the incremental_resolution_regularize(x, y) method to regulate the input resolution based on the dataset being used.
The method applies sub-sampling to the input data to reduce the resolution, depending on the current index and dataset.

Other improvements
Added extra data loaders for some datasets, incremental config files, multiple tests for the algorithm, and a test example for the users to understand. Added necessary checks to ensure that the incremental_grad and incremental_loss_gap flags cannot be used together. Updated comments throughout the code to provide clarity and improve readability.

Looking forward to your feedback and suggestions.

### CODE REVIEW

File: neuralop/training/algo.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            self.resolution = dataset_resolution
+            self.sub_list = dataset_sublist
+
+            self.subsammpling_rate = 1
+            self.current_index = 0
+            self.current_logged_epoch = 0
+            self.current_sub = self.index_to_sub_from_table(self.current_index)
+            self.current_res = self.sub_to_res(self.current_sub)
+
+            print(f'Incre Res Update: change index to {self.current_index}')
```
Comment (JeanKossaifi) [inline neuralop/training/algo.py:44]: Can we add a `verbose` attribute and print only if it's >0 (and also print each update in the corresponding if so it only gets printed when it's actually updated?)

**Review by JeanKossaifi — COMMENTED** [review]
Thanks @Robertboy18 this is much better! Do we need a new trainer or should we merge with the existing one? Also should we test with other models, not just FNO?

File: neuralop/datasets/darcy.py [hunk 2]
```diff
@@ -135,3 +136,80 @@ def load_darcy_pt(data_path,
         test_loaders[res] = test_loader
 
     return train_loader, test_loaders, output_encoder
+
+
+def load_darcy_241(data_path_train, data_path_test,
```
[the remark this answers is not in the document]
Comment (Robertboy18) [reply neuralop/datasets/darcy.py:141]: No we can't do that as this specific training file has different keys compared to the ones in the generic load_darcy. This is the dataset from the original FNO paper so if one wants to use that dataset they have to use this one (see example below the keys are coeff and sol) but one can change the keys and then use the generic load_darcy. Let me know what you think and I can then just discard this function.

```
x_train = torch.tensor(data_train['coeff'], dtype=torch.float)[0:n_train, :, :].unsqueeze(channel_dim).type(torch.float32).clone()
y_train = torch.tensor(data_train['sol'], dtype=torch.float)[0:n_train, :, :].unsqueeze(channel_dim).clone()
```

File: neuralop/models/tests/test_fno_incremental_modes.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    """_summary_
+    Check if the gradients of the non used frequency modes are zero
+    Check if the gradients of the used frequency modes are non zero
+    We only check the corresponding dimensions of the gradients
+    """
+    # set up model
+    modes = (6, 6, 6, 6)
+    incremental_modes = (2, 2, 2, 2)
+    index = list(incremental_modes)
+    for dim in [1, 2, 3, 4]:
```
[the remark this answers is not in the document]
Comment (Robertboy18) [reply neuralop/models/tests/test_fno_incremental_modes.py:45]: Sure!

File: neuralop/training/algo.py (hunk 1 above)
Comment (Robertboy18) [reply neuralop/training/algo.py:44]: Oh I mean initially we want it to print, so we can be sure of which resolution we are starting with before the 1st epoch. Secondly these statements are also only printed when it actually gets updated in the epoch_wise_res_increase function!

File: neuralop/models/tests/test_fno_incremental_modes.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
+"""
+Training just one spectral layer to see if the Incremental Algorithm works - Author Robert Joseph
+"""
+import torch
+import torch.nn as nn
+from neuralop.layers.spectral_convolution import SpectralConv
+
+device = 'cpu'
+
+class Optimizer(torch.optim.SGD):
```
[the remark this answers is not in the document]
Comment (Robertboy18) [reply neuralop/models/tests/test_fno_incremental_modes.py:10]: Cause we want to test if the gradients have been changed, if one just uses this 
```optimizer = torch.optim.SGD(conv.parameters(), lr=0.001) ``` we cannot then get the gradient back by calling it ```            loss, grad = optimizer.step()
``` as optimizer.step() is a None Type. So me and Jiawei decided to just go with this base template and return the gradient data and loss for easier simplicity plus more control over where the calls were made.

File: neuralop/datasets/darcy.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JeanKossaifi) [reply neuralop/datasets/darcy.py:141]: Yes that sounds good - if we can adapt it to match the existing function that would be great

[the remark this answers is not in the document]
Comment (Robertboy18) [reply neuralop/datasets/darcy.py:141]: Okay sure! Yeah, I will just discard this then and add it as a comment then!

File: neuralop/training/algo.py (hunk 1 above)
Comment (Robertboy18) [reply neuralop/training/algo.py]: Sure I can add that!
